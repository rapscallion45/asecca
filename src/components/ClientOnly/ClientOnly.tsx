import { FC, useState, useEffect, ReactNode } from 'react';

interface IClientOnlyProps {
  children: ReactNode;
}

/**
 * Client Only - renders component on client only, no SSR
 *
 * @author - [Carl Scrivener](https://github.com/rapscallion45)
 * @since - 0.0.0
 *
 * @param props - component children
 * @returns {FC} - client only HOC
 */
const ClientOnly: FC<IClientOnlyProps> = (props) => {
  const { children } = props;
  const [hasMounted, setHasMounted] = useState<boolean>(false);

  /** useEffect is only run in the client/browser environment */
  useEffect(() => {
    /** useEffect has run, therefore this is browser/client environment */
    setHasMounted(true);
  }, []);

  return hasMounted ? (
    <div aria-labelledby="client-only-element">{children}</div>
  ) : null;
};

export default ClientOnly;
