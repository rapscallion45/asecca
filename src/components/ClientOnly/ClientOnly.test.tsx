import React, { FC } from 'react';
import { render, screen } from '@testing-library/react';
import renderer from 'react-test-renderer';
import ClientOnly from './ClientOnly';

/**
 * test text
 *
 * @since - 0.0.0
 */
const testMessage = 'Test render client only message';

/**
 * test child component
 *
 * @since - 0.0.0
 */
const TestChild: FC = () => <div>{testMessage}</div>;

/**
 * Client Only Unit Tests
 *
 * @author - [Carl Scrivener](https://github.com/rapscallion45)
 * @since - 0.0.0
 */
describe('Client Only', () => {
  describe('Children', () => {
    it('Should only render if client/browser environment', async () => {
      /** Arrange */
      /** Act */
      render(
        <ClientOnly>
          <TestChild />
        </ClientOnly>
      );

      /** Assert */
      expect(screen.getByText(testMessage)).toBeInTheDocument();
    });
  });

  it('Renders correctly', async () => {
    /** perform snapshot test */
    const tree = renderer
      .create(
        <ClientOnly>
          <TestChild />
        </ClientOnly>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
