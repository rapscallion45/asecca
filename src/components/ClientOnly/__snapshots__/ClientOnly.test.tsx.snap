// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Client Only Renders correctly 1`] = `null`;
